### ISSUE TITLE
Decide on which JUnit version to depend

### ISSUE THREAD

**Issue by nicolaiparlog** [junit-pioneer/junit-pioneer#64]
So far, the project depends on JUnit 5's SNAPSHOT versions. This obviously has to be changed for releases, which begs the question whether we always want to use the latest version? Or depend on the minimum required to work? And what about during development - are snapshots the right choice there? Lots of questions... here are some thoughts:

Depending on snapshots during development...

* ... allows us to develop against unreleased features/APIs
* ... forces us to immediately react to upstream changes
* ... improves JUnit's stability by giving quick feedback from an outside system
* ... requires some work to figure out which version a Pioneer release depends on

Which version of JUnit should released Pioneer artifacts depend on?

* the latest, so we can do the most?
* only update to a newer version when required for important features?
* maybe we have several release lines with different target versions (sounds very complex for a project that doesn't even have any feature, let alone a release)

I have an idea how to go about this, but I'm interested to hear yours first.

**Comment by nicolaiparlog** [comment]
Here are the ideas I alluded to over half a year ago.

## First try

The solution that I find the most compelling is this one:

* during development (i.e. `master` branch) depend on the most recent snapshots
* before each release, determine the minimum required JUnit 5 version and release against that

This gives us the chance to work with the latest and greatest and also gives JUnit 5 an additional round of feedback. There are side effects of course:

* Pioneer might be broken if JUnit 5 breaks ~> great, more feedback for 5!
* Pioneer might be broken until it adapts to breaking upstream changes ~> acceptable as that needs to be done at some point anyway
* new versions of Pioneer are very likely only usable with new versions of 5 ~> since 5 values backwards compatibility, this shouldn't be a big problem
* we might have to release a version (e.g. a bug fix) while we already have code depending on an unreleased version

That last one would hurt a lot because there's no really good fix. Unless we cut release branches. 

## Version proposal

Updated idea:

* `master` depends on the most recent snapshots
* each release line gets its own branch and when a new version gets to be released, we cherry-pick all features and fixes that are compatible with the JUnit version it depends on onto that branch

That begs the question how to cut release branches. This is a separate discussion, so when evaluating  the version proposal, don't conflate the two as all or nothing.

## Branching proposal

Here's my proposal for how to cut release branches:

* name minor releases so that they match the minor JUnit 5 version they depend on (e.g. `1.2.x` ~> `5.2.y`)
* create a branch for each such minor release line

That should be easy to document and intuitive to understand. But I know that @sbrannen recently had a project switch from that approach back to semantic versioning and I'd be interested to hear about the background.

**Comment by marcphilipp** [comment]
I would argue for keeping things simple until there's a real need for making them complicated, i.e. introducing different branches for different versions etc. The JUnit team is committed to keeping things backward compatible so there should be no reason for end users not to use the latest version at all times. However, I can see a point in requiring only the minimum version needed for pioneer. I think that version should be the version that's in the build file.

In addition, I think there should be a CI build that uses the latest snapshot, or possibly a CI matrix with builds for `$minJupiterVersionRequiredByPioneer`, 5.2.x, 5.3.x, ..., `$maxReleasedJupiterVersion`, 5.x.x-SNAPSHOT. There's no need to do that in a branch: a build parameter should suffice.

**Comment by nicolaiparlog** [comment]
> I would argue for keeping things simple until there's a real need for making them complicated

Agreed.

> I think that version should be the version that's in the build file.

But that means no work can be done in Pioneer that depends on unreleased JUnit 5 features / extension points. That qualified as my "real need", but if others don't see it like that, I'm fine with depending on released versions only - surely makes things simpler.

**Comment by marcphilipp** [comment]
I think work that depends on unreleased features should be done in a branch that uses the unreleased version. This way, `master` would always be in a releasable state.

**Comment by nicolaiparlog** [comment]
I considered that, too, and I think it is worth making a full proposal out of it.

## Shadow master

I guess there would be one "shadow master" with all the "unreleasable" features, right? CI would have to build that branch, but that's simple. Then all PRs go onto `master` and as soon as one needs a newer released version of 5 `master` can be upgraded to it.

As soon as a new 5 gets released, all of "shadow master" could be rebased onto `master` and there you go: lots of new features.

I thought I found a problem with this approach (which is why I proposed the one above), but for the life of me, I can't came up with is, so for now, this is my favorite approach.

I would like to hear @smoyer64's opinion on this before making a decision, though.

**Comment by sbrannen** [comment]
I agree with @marcphilipp: I don't see any need to make things unnecessarily complicated just for the sake of it.

Let's just stick with `master` using the latest GA releases of JUnit 5.x, have a separate build against latest snapshots (as a "heads up" for potentially upcoming incompatibilities), and only build new features against snapshots if absolutely necessary (i.e., in feature dedicated feature branches).

In other words, let's hold off on introducing an official "shadow master" until a concrete need arises.

my $0.02,

Sam

**Comment by smoyer64** [comment]
@nicolaiparlog If I'm reading your "Shadow Master" paragraph correctly, that would only be required for new features that rely on Jupiter features that haven't been released yet right?

I'm far enough behind that I just need to get the extensions that I have moved over into Pioneer ... do you have any features that rely on > 5.2.0 at the moment?  (I have great plans for ScenariosTheories/Ordered tests for UI testing but no ideas for extensions to any of them yet).
